File: Liftr/Chat/ChatModels.swift
import Foundation
import Supabase

enum ChatKind: String, Decodable {
    case direct
    case group
}

enum ChatMessageKind: String, Codable {
    case text
    case image
    case file
    case system
    case workoutShare      = "workout_share"
    case routineShare      = "routine_share"
    case achievementShare  = "achievement_share"
    case segmentShare      = "segment_share"
}

struct ConversationOverview: Decodable, Identifiable, Hashable {
    let id: Int64
    let kind: String
    let title: String?
    let updated_at: Date
    let last_message_id: Int64?
    let last_message_user_id: UUID?
    let last_message_body: String?
    let last_message_at: Date?
    let unread_count: Int

    var conversationKind: ChatKind { ChatKind(rawValue: kind) ?? .direct }
}

struct ChatMessage: Decodable, Identifiable, Hashable {
    let id: Int64
    let user_id: UUID
    let kind: String
    let body: String?
    let metadata: AnyJSON?
    let reply_to_message_id: Int64?
    let created_at: Date
    let edited_at: Date?
    let deleted_at: Date?

    var messageKind: ChatMessageKind { ChatMessageKind(rawValue: kind) ?? .text }

    init(id: Int64,
         user_id: UUID,
         kind: String,
         body: String?,
         metadata: AnyJSON?,
         reply_to_message_id: Int64? = nil,
         created_at: Date,
         edited_at: Date? = nil,
         deleted_at: Date? = nil) {
        self.id = id
        self.user_id = user_id
        self.kind = kind
        self.body = body
        self.metadata = metadata
        self.reply_to_message_id = reply_to_message_id
        self.created_at = created_at
        self.edited_at = edited_at
        self.deleted_at = deleted_at
    }

    private enum CodingKeys: String, CodingKey {
        case id, user_id, kind, body, metadata, reply_to_message_id
        case created_at, edited_at, deleted_at
    }

    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        self.id = try c.decode(Int64.self, forKey: .id)
        self.user_id = try c.decode(UUID.self, forKey: .user_id)
        self.kind = try c.decode(String.self, forKey: .kind)
        self.body = try c.decodeIfPresent(String.self, forKey: .body)
        self.metadata = try c.decodeIfPresent(AnyJSON.self, forKey: .metadata)
        self.reply_to_message_id = try c.decodeIfPresent(Int64.self, forKey: .reply_to_message_id)
        self.created_at = try c.decode(Date.self, forKey: .created_at)
        self.edited_at = try c.decodeIfPresent(Date.self, forKey: .edited_at)
        self.deleted_at = try c.decodeIfPresent(Date.self, forKey: .deleted_at)
    }
}

enum ReactionEmoji: String, CaseIterable, Codable, Hashable {
    case heart      = "heart"
    case haha       = "haha"
    case wow        = "wow"
    case sad        = "sad"
    case thumbsUp   = "thumbs_up"
    case thumbsDown = "thumbs_down"

    var glyph: String {
        switch self {
        case .heart:      return "❤️"
        case .haha:       return "😂"
        case .wow:        return "😮"
        case .sad:        return "😢"
        case .thumbsUp:   return "👍"
        case .thumbsDown: return "👎"
        }
    }
}

struct MessageReaction: Decodable, Hashable {
    let message_id: Int64
    let user_id: UUID
    let emoji: String
    let created_at: Date

    var reaction: ReactionEmoji? { ReactionEmoji(rawValue: emoji) }
}

struct ChatProfile: Decodable, Identifiable, Hashable {
    let user_id: UUID
    let username: String
    let avatar_url: String?
    var id: UUID { user_id }
}

struct WorkoutShareSnapshot: Codable, Hashable {
    let v: Int
    let workout_id: Int64
    let title: String?
    let kind: String?
    let score: Int?
    let kcal: Int?
    let performed_at: String?
    let owner_user_id: UUID?
    let owner_username: String?
    let owner_avatar_url: String?
}

struct RoutineShareSnapshot: Codable, Hashable {
    let v: Int
    let type: String
    let routine_kind: String
    let name: String
    let routine_id: Int64?
    let updated_at: String?
    let owner_user_id: UUID?
    let owner_username: String?
    let owner_avatar_url: String?
    let share_nonce: String
    let detail_json: String
    let exercise_count: Int?
    let total_sets: Int?
    let preview_exercise_name: String?

    init(v: Int,
         type: String,
         routine_kind: String,
         name: String,
         routine_id: Int64?,
         updated_at: String?,
         owner_user_id: UUID?,
         owner_username: String?,
         owner_avatar_url: String?,
         share_nonce: String,
         detail_json: String,
         exercise_count: Int? = nil,
         total_sets: Int? = nil,
         preview_exercise_name: String? = nil) {
        self.v = v
        self.type = type
        self.routine_kind = routine_kind
        self.name = name
        self.routine_id = routine_id
        self.updated_at = updated_at
        self.owner_user_id = owner_user_id
        self.owner_username = owner_username
        self.owner_avatar_url = owner_avatar_url
        self.share_nonce = share_nonce
        self.detail_json = detail_json
        self.exercise_count = exercise_count
        self.total_sets = total_sets
        self.preview_exercise_name = preview_exercise_name
    }
}

struct AchievementShareSnapshot: Codable, Hashable {
    let v: Int
    let type: String
    let code: String
    let achievement_id: Int
    let title: String
    let category: String
    let description: String?
    let icon_url: String?
    let owner_user_id: UUID?
    let owner_username: String?
    let owner_avatar_url: String?
}

struct SegmentShareSnapshot: Codable, Hashable {
    let v: Int
    let type: String
    let segment_id: String
    let name: String
    let segment_length_m: Double?
    let leaderboard_effort_count: Int64?
    let owner_user_id: UUID?
    let owner_username: String?
    let owner_avatar_url: String?
}

extension ChatMessage {
    func workoutShare() -> WorkoutShareSnapshot? {
        guard messageKind == .workoutShare, let metadata else { return nil }
        return try? metadata.decode(as: WorkoutShareSnapshot.self)
    }

    func routineShare() -> RoutineShareSnapshot? {
        guard messageKind == .routineShare, let metadata else { return nil }
        return try? metadata.decode(as: RoutineShareSnapshot.self)
    }

    func achievementShare() -> AchievementShareSnapshot? {
        guard messageKind == .achievementShare, let metadata else { return nil }
        return try? metadata.decode(as: AchievementShareSnapshot.self)
    }

    func segmentShare() -> SegmentShareSnapshot? {
        guard messageKind == .segmentShare, let metadata else { return nil }
        return try? metadata.decode(as: SegmentShareSnapshot.self)
    }
}

File: android/app/src/main/java/com/lilru/liftr/ui/chat/ChatModels.kt
package com.lilru.liftr.ui.chat

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.decodeFromJsonElement

/**
 * Inbox row produced by `public.get_conversations_overview()`.
 */
@Serializable
data class ConversationOverviewWire(
    val id: Long,
    val kind: String,
    val title: String? = null,
    @SerialName("updated_at") val updatedAt: String,
    @SerialName("last_message_id") val lastMessageId: Long? = null,
    @SerialName("last_message_user_id") val lastMessageUserId: String? = null,
    @SerialName("last_message_body") val lastMessageBody: String? = null,
    @SerialName("last_message_at") val lastMessageAt: String? = null,
    @SerialName("unread_count") val unreadCount: Int = 0
)

/**
 * Single message returned by `public.get_messages()` and broadcast events.
 */
@Serializable
data class ChatMessageWire(
    val id: Long,
    @SerialName("user_id") val userId: String,
    val kind: String = "text",
    val body: String? = null,
    val metadata: JsonObject? = null,
    @SerialName("reply_to_message_id") val replyToMessageId: Long? = null,
    @SerialName("created_at") val createdAt: String,
    @SerialName("edited_at") val editedAt: String? = null,
    @SerialName("deleted_at") val deletedAt: String? = null,
    @SerialName("conversation_id") val conversationId: Long? = null
)

/**
 * One reaction row produced by `public.message_reactions` (and the
 * `conversation:<id>:reactions` realtime topic).
 */
@Serializable
data class ReactionWire(
    @SerialName("message_id") val messageId: Long,
    @SerialName("user_id") val userId: String,
    val emoji: String,
    @SerialName("created_at") val createdAt: String
)

/**
 * Six-emoji set matching the CHECK constraint on
 * `public.message_reactions.emoji`.
 */
enum class ReactionEmoji(val raw: String, val glyph: String) {
    HEART("heart", "\u2764\ufe0f"),
    HAHA("haha", "\ud83d\ude02"),
    WOW("wow", "\ud83d\ude2e"),
    SAD("sad", "\ud83d\ude22"),
    THUMBS_UP("thumbs_up", "\ud83d\udc4d"),
    THUMBS_DOWN("thumbs_down", "\ud83d\udc4e");

    companion object {
        fun fromRaw(raw: String): ReactionEmoji? = entries.firstOrNull { it.raw == raw }
    }
}

/**
 * Compact preview of a parent message used by the in-bubble reply chip.
 */
@Serializable
data class ReplyPreviewWire(
    val id: Long,
    @SerialName("user_id") val userId: String,
    val body: String? = null,
    val kind: String? = null,
    val metadata: JsonObject? = null,
    @SerialName("deleted_at") val deletedAt: String? = null
)

/**
 * Row from `conversation_reads` used to build the "Seen" indicator.
 */
@Serializable
data class ConversationReadRow(
    @SerialName("user_id") val userId: String,
    @SerialName("last_read_message_id") val lastReadMessageId: Long? = null
)

/**
 * Mute flag on `conversation_participants` for the current user.
 */
@Serializable
data class ParticipantMutedRow(
    val muted: Boolean = false
)

@Serializable
data class ProfileLite(
    @SerialName("user_id") val userId: String,
    val username: String,
    @SerialName("avatar_url") val avatarUrl: String? = null
)

@Serializable
data class ChatParticipantRow(
    @SerialName("conversation_id") val conversationId: Long,
    @SerialName("user_id") val userId: String
)

@Serializable
data class FollowEdge(
    @SerialName("follower_id") val followerId: String? = null,
    @SerialName("followee_id") val followeeId: String? = null
)

/**
 * `messages.kind` values understood by the client. New entries (e.g. share
 * variants) must default to [TEXT] in older builds via [fromRaw], so unknown
 * kinds keep rendering as plain bubbles instead of crashing.
 */
enum class ChatKind(val raw: String) {
    TEXT("text"),
    IMAGE("image"),
    FILE("file"),
    SYSTEM("system"),
    WORKOUT_SHARE("workout_share"),
    ROUTINE_SHARE("routine_share"),
    ACHIEVEMENT_SHARE("achievement_share"),
    SEGMENT_SHARE("segment_share"),
    SHARED_INGREDIENT("shared_ingredient"),
    SHARED_RECIPE("shared_recipe");

    companion object {
        fun fromRaw(raw: String?): ChatKind =
            entries.firstOrNull { it.raw == raw } ?: TEXT
    }
}

/**
 * Compact, self-contained snapshot of a workout that travels inside a
 * `workout_share` message's `metadata`. The snapshot keeps the chat preview
 * readable even if the original workout is later deleted.
 */
@Serializable
data class WorkoutShareSnapshot(
    val v: Int = 1,
    val type: String = "workout_share",
    @SerialName("workout_id") val workoutId: Long,
    val title: String? = null,
    val kind: String? = null,
    val score: Int? = null,
    val kcal: Int? = null,
    @SerialName("performed_at") val performedAt: String? = null,
    @SerialName("owner_user_id") val ownerUserId: String? = null,
    @SerialName("owner_username") val ownerUsername: String? = null,
    @SerialName("owner_avatar_url") val ownerAvatarUrl: String? = null
)

@Serializable
data class RoutineShareSnapshot(
    val v: Int = 1,
    val type: String = "routine_share",
    @SerialName("routine_kind") val routineKind: String,
    val name: String,
    @SerialName("routine_id") val routineId: Long? = null,
    @SerialName("updated_at") val updatedAt: String? = null,
    @SerialName("owner_user_id") val ownerUserId: String? = null,
    @SerialName("owner_username") val ownerUsername: String? = null,
    @SerialName("owner_avatar_url") val ownerAvatarUrl: String? = null,
    @SerialName("share_nonce") val shareNonce: String,
    @SerialName("detail_json") val detailJson: String,
    @SerialName("exercise_count") val exerciseCount: Int? = null,
    @SerialName("total_sets") val totalSets: Int? = null,
    @SerialName("preview_exercise_name") val previewExerciseName: String? = null
)

@Serializable
data class AchievementShareSnapshot(
    val v: Int = 1,
    val type: String = "achievement_share",
    val code: String,
    @SerialName("achievement_id") val achievementId: Int,
    val title: String,
    val category: String,
    val description: String? = null,
    @SerialName("icon_url") val iconUrl: String? = null,
    @SerialName("owner_user_id") val ownerUserId: String? = null,
    @SerialName("owner_username") val ownerUsername: String? = null,
    @SerialName("owner_avatar_url") val ownerAvatarUrl: String? = null
)

@Serializable
data class SegmentShareSnapshot(
    val v: Int = 1,
    val type: String = "segment_share",
    @SerialName("segment_id") val segmentId: String,
    val name: String,
    @SerialName("segment_length_m") val segmentLengthM: Double? = null,
    @SerialName("leaderboard_effort_count") val leaderboardEffortCount: Long? = null,
    @SerialName("owner_user_id") val ownerUserId: String? = null,
    @SerialName("owner_username") val ownerUsername: String? = null,
    @SerialName("owner_avatar_url") val ownerAvatarUrl: String? = null
)

@Serializable
data class SharedIngredientSnapshot(
    val v: Int = 1,
    val type: String = "shared_ingredient",
    val name: String,
    @SerialName("calories_per_100g") val caloriesPer100g: Double,
    @SerialName("protein_per_100g") val proteinPer100g: Double = 0.0,
    @SerialName("carbs_per_100g") val carbsPer100g: Double = 0.0,
    @SerialName("fat_per_100g") val fatPer100g: Double = 0.0,
    @SerialName("saturated_fat_per_100g") val saturatedFatPer100g: Double = 0.0,
    @SerialName("sugars_per_100g") val sugarsPer100g: Double = 0.0,
    @SerialName("fiber_per_100g") val fiberPer100g: Double = 0.0,
    @SerialName("sodium_mg_per_100g") val sodiumMgPer100g: Double = 0.0
)

@Serializable
data class SharedRecipeIngredientSnapshot(
    val name: String,
    @SerialName("weight_g") val weightG: Double,
    @SerialName("calories_per_100g") val caloriesPer100g: Double,
    @SerialName("protein_per_100g") val proteinPer100g: Double = 0.0,
    @SerialName("carbs_per_100g") val carbsPer100g: Double = 0.0,
    @SerialName("fat_per_100g") val fatPer100g: Double = 0.0,
    @SerialName("saturated_fat_per_100g") val saturatedFatPer100g: Double = 0.0,
    @SerialName("sugars_per_100g") val sugarsPer100g: Double = 0.0,
    @SerialName("fiber_per_100g") val fiberPer100g: Double = 0.0,
    @SerialName("sodium_mg_per_100g") val sodiumMgPer100g: Double = 0.0
)

@Serializable
data class SharedRecipeProfilePer100gSnapshot(
    val calories: Double,
    val protein: Double,
    val carbs: Double,
    val fat: Double,
    val saturatedFat: Double,
    val sugars: Double,
    val fiber: Double,
    val sodiumMg: Double
)

@Serializable
data class SharedRecipeSnapshot(
    val v: Int = 1,
    val type: String = "shared_recipe",
    val name: String,
    val description: String? = null,
    val ingredients: List<SharedRecipeIngredientSnapshot>,
    @SerialName("profile_per_100g") val profilePer100g: SharedRecipeProfilePer100gSnapshot? = null
)

private val shareMetadataJson = Json { ignoreUnknownKeys = true }

private fun JsonObject.withDetailJsonCoercedToString(): JsonObject {
    return buildJsonObject {
        this@withDetailJsonCoercedToString.forEach { (k, v) ->
            if (k == "detail_json") {
                val normalized = when (v) {
                    is JsonPrimitive -> v
                    is JsonObject -> JsonPrimitive(v.toString())
                    is JsonArray -> JsonPrimitive(v.toString())
                    else -> v
                }
                put(k, normalized)
            } else {
                put(k, v)
            }
        }
    }
}

fun ChatMessageWire.decodeWorkoutShare(): WorkoutShareSnapshot? {
    if (ChatKind.fromRaw(kind) != ChatKind.WORKOUT_SHARE) return null
    val m = metadata ?: return null
    return runCatching { shareMetadataJson.decodeFromJsonElement(WorkoutShareSnapshot.serializer(), m) }.getOrNull()
}

fun ChatMessageWire.decodeRoutineShare(): RoutineShareSnapshot? {
    if (ChatKind.fromRaw(kind) != ChatKind.ROUTINE_SHARE) return null
    val m = metadata ?: return null
    val normalized = m.withDetailJsonCoercedToString()
    return runCatching {
        shareMetadataJson.decodeFromJsonElement(RoutineShareSnapshot.serializer(), normalized)
    }.getOrNull()
}

fun ChatMessageWire.decodeAchievementShare(): AchievementShareSnapshot? {
    if (ChatKind.fromRaw(kind) != ChatKind.ACHIEVEMENT_SHARE) return null
    val m = metadata ?: return null
    return runCatching {
        shareMetadataJson.decodeFromJsonElement(AchievementShareSnapshot.serializer(), m)
    }.getOrNull()
}

fun ChatMessageWire.decodeSegmentShare(): SegmentShareSnapshot? {
    if (ChatKind.fromRaw(kind) != ChatKind.SEGMENT_SHARE) return null
    val m = metadata ?: return null
    return runCatching {
        shareMetadataJson.decodeFromJsonElement(SegmentShareSnapshot.serializer(), m)
    }.getOrNull()
}

fun ChatMessageWire.decodeSharedIngredient(): SharedIngredientSnapshot? {
    if (ChatKind.fromRaw(kind) != ChatKind.SHARED_INGREDIENT) return null
    val m = metadata ?: return null
    return runCatching {
        shareMetadataJson.decodeFromJsonElement(SharedIngredientSnapshot.serializer(), m)
    }.getOrNull()
}

fun ChatMessageWire.decodeSharedRecipe(): SharedRecipeSnapshot? {
    if (ChatKind.fromRaw(kind) != ChatKind.SHARED_RECIPE) return null
    val m = metadata ?: return null
    return runCatching {
        shareMetadataJson.decodeFromJsonElement(SharedRecipeSnapshot.serializer(), m)
    }.getOrNull()
}

fun ChatMessageWire.replyComposerSubtitle(): String = when (ChatKind.fromRaw(kind)) {
    ChatKind.ROUTINE_SHARE -> decodeRoutineShare()?.let { r -> "Routine · ${r.name}" } ?: body.orEmpty()
    ChatKind.WORKOUT_SHARE -> decodeWorkoutShare()?.let { w ->
        val t = w.title?.trim()?.takeIf { it.isNotEmpty() }
        if (t != null) "Workout · $t" else "Workout"
    } ?: body.orEmpty()
    ChatKind.ACHIEVEMENT_SHARE -> decodeAchievementShare()?.let { a ->
        val t = a.title.trim().takeIf { it.isNotEmpty() }
        if (t != null) "Achievement · $t" else "Achievement"
    } ?: body.orEmpty()
    ChatKind.SEGMENT_SHARE -> decodeSegmentShare()?.let { s ->
        val t = s.name.trim().takeIf { it.isNotEmpty() }
        if (t != null) "Segment · $t" else "Segment"
    } ?: body.orEmpty()
    ChatKind.SHARED_INGREDIENT -> decodeSharedIngredient()?.let { "Ingredient · ${it.name}" } ?: body.orEmpty()
    ChatKind.SHARED_RECIPE -> decodeSharedRecipe()?.let { "Recipe · ${it.name}" } ?: body.orEmpty()
    else -> body.orEmpty()
}

fun ChatMessageWire.clipboardTextForCopy(): String = when (ChatKind.fromRaw(kind)) {
    ChatKind.ROUTINE_SHARE -> decodeRoutineShare()?.let { r ->
        buildString {
            append("Routine: ${r.name}")
            r.ownerUsername?.takeIf { it.isNotBlank() }?.let { append("\nFrom @").append(it) }
        }
    } ?: body.orEmpty()
    ChatKind.WORKOUT_SHARE -> decodeWorkoutShare()?.let { w ->
        buildString {
            append("Workout")
            w.title?.takeIf { it.isNotBlank() }?.let { append(": ").append(it) }
            w.ownerUsername?.takeIf { it.isNotBlank() }?.let { append("\nFrom @").append(it) }
        }
    } ?: body.orEmpty()
    ChatKind.ACHIEVEMENT_SHARE -> decodeAchievementShare()?.let { a ->
        buildString {
            append("Achievement: ${a.title}")
            a.ownerUsername?.takeIf { it.isNotBlank() }?.let { append("\nFrom @").append(it) }
        }
    } ?: body.orEmpty()
    ChatKind.SEGMENT_SHARE -> decodeSegmentShare()?.let { s ->
        buildString {
            append("Segment: ${s.name}")
            s.ownerUsername?.takeIf { it.isNotBlank() }?.let { append("\nFrom @").append(it) }
        }
    } ?: body.orEmpty()
    ChatKind.SHARED_INGREDIENT -> decodeSharedIngredient()?.let { s ->
        buildString { append("Ingredient: ${s.name}") }
    } ?: body.orEmpty()
    ChatKind.SHARED_RECIPE -> decodeSharedRecipe()?.let { s ->
        buildString { append("Recipe: ${s.name}") }
    } ?: body.orEmpty()
    else -> body.orEmpty()
}

fun ReplyPreviewWire.previewText(): String {
    if (deletedAt != null) return "Original message was deleted"
    return when (ChatKind.fromRaw(kind)) {
        ChatKind.ROUTINE_SHARE -> {
            val m = metadata ?: return body.orEmpty()
            runCatching {
                shareMetadataJson.decodeFromJsonElement(
                    RoutineShareSnapshot.serializer(),
                    m.withDetailJsonCoercedToString()
                ).name
            }.getOrNull()?.let { "Routine · $it" } ?: body.orEmpty()
        }
        ChatKind.WORKOUT_SHARE -> {
            val m = metadata ?: return body.orEmpty()
            runCatching {
                shareMetadataJson.decodeFromJsonElement(WorkoutShareSnapshot.serializer(), m)
            }.getOrNull()?.let { w ->
                val t = w.title?.trim()?.takeIf { it.isNotEmpty() }
                if (t != null) "Workout · $t" else "Workout"
            } ?: body.orEmpty()
        }
        ChatKind.ACHIEVEMENT_SHARE -> {
            val m = metadata ?: return body.orEmpty()
            runCatching {
                shareMetadataJson.decodeFromJsonElement(AchievementShareSnapshot.serializer(), m)
            }.getOrNull()?.let { a ->
                val t = a.title.trim().takeIf { it.isNotEmpty() }
                if (t != null) "Achievement · $t" else "Achievement"
            } ?: body.orEmpty()
        }
        ChatKind.SEGMENT_SHARE -> {
            val m = metadata ?: return body.orEmpty()
            runCatching {
                shareMetadataJson.decodeFromJsonElement(SegmentShareSnapshot.serializer(), m)
            }.getOrNull()?.let { s ->
                val t = s.name.trim().takeIf { it.isNotEmpty() }
                if (t != null) "Segment · $t" else "Segment"
            } ?: body.orEmpty()
        }
        ChatKind.SHARED_INGREDIENT -> {
            val m = metadata ?: return body.orEmpty()
            runCatching {
                shareMetadataJson.decodeFromJsonElement(SharedIngredientSnapshot.serializer(), m)
            }.getOrNull()?.let { s ->
                val t = s.name.trim().takeIf { it.isNotEmpty() }
                if (t != null) "Ingredient · $t" else "Ingredient"
            } ?: body.orEmpty()
        }
        ChatKind.SHARED_RECIPE -> {
            val m = metadata ?: return body.orEmpty()
            runCatching {
                shareMetadataJson.decodeFromJsonElement(SharedRecipeSnapshot.serializer(), m)
            }.getOrNull()?.let { s ->
                val t = s.name.trim().takeIf { it.isNotEmpty() }
                if (t != null) "Recipe · $t" else "Recipe"
            } ?: body.orEmpty()
        }
        else -> body.orEmpty()
    }
}
